### DISCUSSION TITLE
Is there a way for multiple consumers?

### DISCUSSION THREAD

**Discussion by bennyKY** [taskforcesh/bullmq#400] [Q&A]
Hi, Bull uses Redis pubsub so it wont allow multiple consumers.
But Bullmq uses Redis streams so it should be feasible.
Is there a way to do it with Bullmq? If not maybe someone can advise how to achieve it without Rabbit or Kafka?

**Comment by manast** [comment]
Bull and BullMQ both support multiple consumers, just create as many workers as you see fit and they will consume your jobs in parallell.

**Comment by bennyKY** [comment]
@manast 
I think i wasnt clear, sorry for that. Im talking about broadcasting. Multiple services that consume from a queue. As far as i worked with Bull (1 year in production) if 2 services listening to messages from a queue only 1 services will receive the message.

**Comment by manast** [comment]
Ok, so you mean fanout (https://en.wikipedia.org/wiki/Fan-out_(software)), i.e. same job to many consumers. That is not supported and there are no plans for it, in that case using XTREAMS is a really good solution if you do not want to experiment with more complex solutions such as Kafka.

**Comment by bennyKY** [comment]
@manast thanks a lot for your answers, sorry didnt find anything related to XTREAMS, can you please give me a reference?

**Comment by manast** [comment]
@bennyKY I meant streams as you wrote on the issue, sorry for the confusion: https://redis.io/commands#stream

**Comment by audenaert** [comment]
@manast Came across this while searching for the same goal as the OP.

It looks like I could achieve a similar result by listening to queue events. Seems like a bit of an abuse of the intention of BullMQ, but wondering if something along the following lines would get a similar result.

```Javascript
const queue = new Queue('Cars');
await queue.add('paint', { color: 'red' });

const worker = new Worker('paint', paintHandler);
...
const queueEvents = new QueueEvents('Paint');
queueEvents.on('completed', queueListener1);
queueEvents.on('completed', queueListener2);
queueEvents.on('completed', queueListener3);
```

The downside here (I assume) is that I can't set these up to run in multiple workers because each listener would be called in each worker. That could lead to surprising behavior unless my listeners are idempotent. For example, if the app that's listening scales horizontally, each listener would be called on each running instance.

Am I understanding how Queue listeners work correctly?

**Reply by notflip** [reply]
@audenaert Did you ever find an answer to your question?

**Reply by manast** [reply]
The answer to the question is that there will be no surprising behaviour here, the job will be processed by one of the workers and all the global listeners (queueEvents instances), listening to the 'completed' event will receive the event.

**Comment by adamlamaa** [comment]
Had to eliminate BullMQ because it does not provide fanout, looking at RabbitMQ or LavinMQ instead.

**Reply by manast** [reply]
You can achieve fanout quite easily using Redis pubsub...

**Reply by manast** [reply]
The thing with fanout is that there are actually 2 ways to handle it. Like Redis pubsub, which basically is a broadcast mechanism, some producer puts a message on a channel, and then all the consumers that are listening to that channel at that point in time will receive the message. However, this approach implies that the listeners that are not online at the moment of the publish will not receive the message. This is better handled by Redis streams, in that you can send a message to a stream, and all the listeners can receive them as long as they keep a last-message-id for that stream, so that they can always go back in time and be sure that they are getting all the messages. Both patterns can be done with Redis directly pretty easily, no need for a queue broker for this unless I am missing something...

**Reply by adamlamaa** [reply]
Not sure if you are suggesting Redis pubsub as an alternative or as a way to acheive fanout with BullMQ @manast ?
For info I tried setting up BullMQ in a NestJS server, I was not able to find a way to acheive a fanout setup. If I have more then one @ Process tag on the same queue I get this error:

```
.../node_modules/bull/lib/queue.js:676
    throw new Error('Cannot define the same handler twice ' + name);
```
    
Was not able to find any references/guides to fanout or pubsub in the BullMQ docs.

**Reply by manast** [reply]
Can you elaborate what do you mean with "fanout" ?

**Reply by adamlamaa** [reply]
Did not see your detailed reply before my follow up question. Thanks for sharing your perspective. I will have a look at just using Redis on its own.

**Reply by manast** [reply]
Btw, you should probably use the BullMQ version instead of bull: https://docs.bullmq.io/guide/nestjs

**Reply by adamlamaa** [reply]
The architecture I am looking to implement is the ability to broadcast an event, which can then be picked up by multiple subscribers. Each subscriber can then evaluate on their own if they need to perform any actions as a result of the event.

Enabling a microservice setup, where logic can be split up and the code be structured in a more event driven way.

It was my understanding that in the queue world the pattern that fits this need is fanout.

And yes, was using the nestjs version, thanks!

**Reply by adamlamaa** [reply]
Update: I was not able to acheive the dream setup I had invisioned in my head without any MQ.

Ended up using BullMQ with a FlowProducer.

**Comment by yisibl** [comment]
Is this a needed feature for this issue? https://github.com/taskforcesh/bullmq/pull/2854

**Reply by manast** [reply]
For fanout you could also use custom global events, not sure if it would fullfil the requirements for a pubsub scenario but seems to be quite close: https://docs.bullmq.io/guide/events/create-custom-events
